transport/src/main/java/io/netty/channel/ManualIoEventLoop.java

```diff
@@ -148,28 +149,71 @@ public Ticker ticker() {
         return ticker;
     }
 
-    private int runAllTasks() {
+    /**
+     * Poll all tasks from the task queue and run them via {@link Runnable#run()} method.
+     * <br>
+     * The behaviour of this method is the same as {@link #runNonIoTasks(long)}
+     * with {@code timeoutNanos} set {@code <= 0}.
+     */
+    public int runNonIoTasks() {
+        return runAllTasks(-1, true);
+    }
+
+    /**
+     * Poll all tasks from the task queue and run them via {@link Runnable#run()} method.  This method stops running
+     * the tasks in the task queue and returns if it ran longer than {@code timeoutNanos}.<br>
+     *
+     * @param timeoutNanos the maximum time in nanoseconds to run tasks. If {@code <= 0}, no timeout is applied.
+     */
+    public int runNonIoTasks(long timeoutNanos) {
+        return runAllTasks(timeoutNanos, true);
+    }
+
+    private int runAllTasks(long timeoutNanos, boolean setCurrentExecutor) {
         assert inEventLoop();
-        int numRun = 0;
-        boolean fetchedAll;
-        do {
-            fetchedAll = fetchFromScheduledTaskQueue(taskQueue);
+        final Queue<Runnable> taskQueue = this.taskQueue;
+        // since the taskQueue is unbounded we don't need to keep on calling this while draining it.
+        boolean alwaysTrue = fetchFromScheduledTaskQueue(taskQueue);
+        assert alwaysTrue;
+        Runnable task = taskQueue.poll();
+        if (task == null) {
+            return 0;
+        }
+        EventExecutor old = setCurrentExecutor? ThreadExecutorMap.setCurrentExecutor(this) : null;
+        try {
+            final long deadline = timeoutNanos > 0 ? getCurrentTimeNanos() + timeoutNanos : 0;
+            int runTasks = 0;
+            long lastExecutionTime;
+            final Ticker ticker = this.ticker;
             for (;;) {
-                Runnable task = taskQueue.poll();
+                safeExecute(task);
+
+                runTasks++;
+
+                // Check timeout every each TICKER_FREQUENCY tasks because nanoTime() is relatively expensive.
```

Comment:
Comment look obsolete? I don't see any TICKER_FREQUENCY, and here we call the ticker each iteration (which I think is correct), which will usually go to System.nanoTime().